### PR TITLE
Refactor `StoreEvent` and create a new `RoomDatabase` interface

### PR DESCRIPTION
**Pull request by S7evinK**
This PR changes a few things:
- It pulls out the creation of several NIDs from the `StoreEvent` function to make the functions more reusable
- Uses more caching when using those NIDs to avoid DB round trips

### CODE REVIEW

**Review by kegsay — APPROVED** [review]
This is a good start for unpicking room stuff from the roomserver.

File: roomserver/internal/input/input_events.go [hunk 1]
```diff
@@ -308,10 +308,10 @@ func (r *Inputer) processRoomEvent(
 	}
 
 	var softfail bool
-	if input.Kind == api.KindNew {
+	if input.Kind == api.KindNew && !isCreateEvent {
```
[the remark this answers is not in the document]
Comment (S7evinK) [reply roomserver/internal/input/input_events.go:311]: A create event shouldn't be able to soft-fail and we wouldn't have a `roomInfo` so would run into nil pointer issues.

File: roomserver/internal/input/input_events.go [hunk 2]
```diff
@@ -332,8 +332,23 @@ func (r *Inputer) processRoomEvent(
...
+	if err != nil {
+		return fmt.Errorf("r.DB.GetOrCreateRoomNID: %w", err)
+	}
+
+	eventTypeNID, err := r.DB.GetOrCreateEventTypeNID(ctx, event.Type())
+	if err != nil {
+		return fmt.Errorf("r.DB.GetOrCreateEventTypeNID: %w", err)
+	}
+
+	eventStateKeyNID, err := r.DB.GetOrCreateEventStateKeyNID(ctx, event.StateKey())
```
[the remark this answers is not in the document]
Comment (S7evinK) [reply roomserver/internal/input/input_events.go:345]: Yep, probably something for after the refactoring.

File: roomserver/storage/shared/storage.go [hunk 3]
```diff
@@ -1470,13 +1509,19 @@ func (d *Database) PurgeRoom(ctx context.Context, roomID string) error {
...
-		if err := d.PublishedTable.UpsertRoomPublished(ctx, txn, oldRoomID, "", "", false); err != nil {
-			return fmt.Errorf("failed to unpublish room: %w", err)
+		published, err := d.PublishedTable.SelectPublishedFromRoomID(ctx, txn, oldRoomID)
+		if err != nil {
+			return fmt.Errorf("failed to get published room: %w", err)
 		}
-		// publish new room
-		if err := d.PublishedTable.UpsertRoomPublished(ctx, txn, newRoomID, "", "", true); err != nil {
-			return fmt.Errorf("failed to publish room: %w", err)
+		if published {
```
[the remark this answers is not in the document]
Comment (S7evinK) [reply roomserver/storage/shared/storage.go:1516]: Falsely assumed that every room is published, which is not the case. So this fixes "automatic publishing rooms on upgrade"

**Pull request merged**